### PR TITLE
[5.5] Crypto-based password resets

### PR DESCRIPTION
**Pull request by tillkruss**
Same as https://github.com/laravel/framework/pull/17052.

As discussed with @taylorotwell, a proof-of-concept for crypto-based password resets.

This approach allows us to drop the `password_resets` table and remove the `TokenRepositoryInterface` and `DatabaseTokenRepository`. See `laravel/laravel` [changes.](https://github.com/laravel/laravel/compare/master...tillkruss:passwords)

The email field has been removed from the reset form.

If the reset link is expired/invalid, or the user doesn't exists, then the password reset form is hidden and an appropriate warning message is displayed.

### CODE REVIEW

**Comment by juukie** [comment]
👏 nice

File: src/Illuminate/Auth/Passwords/PasswordBroker.php [hunk 1]
```diff
@@ -200,43 +213,72 @@ public function getUser(array $credentials)
...
+     * Return the application key.
+     *
+     * @return string
+     */
+    public function getKey()
     {
-        return $this->tokens->exists($user, $token);
+        $key = $this->app['config']['app.key'];
+
+        if (Str::startsWith($key, 'base64:')) {
```
Comment (garygreen) [inline src/Illuminate/Auth/Passwords/PasswordBroker.php]: Does nowhere else in the framework do this base64 decoding of the app key? Seems like there should be...

**Comment by garygreen** [comment]
Awesome stuff! 👍

File: src/Illuminate/Auth/Passwords/PasswordBroker.php (hunk 1 above)
Comment (tillkruss) [inline src/Illuminate/Auth/Passwords/PasswordBroker.php]: Checkout `EncryptionServiceProvider` and `KeyGenerateCommand`.

Comment (garygreen) [inline src/Illuminate/Auth/Passwords/PasswordBroker.php]: Ah, well that kind of code duplication is getting a bit messy. Going off topic a bit but I've proposed some improvements to this laravel/internals/issues/416

File: src/Illuminate/Auth/Passwords/PasswordBroker.php [hunk 2]
```diff
@@ -35,15 +45,15 @@ class PasswordBroker implements PasswordBrokerContract
     /**
      * Create a new password broker instance.
      *
-     * @param  \Illuminate\Auth\Passwords\TokenRepositoryInterface  $tokens
+     * @param  \Illuminate\Contracts\Foundation\Application  $app
```
Comment (taylorotwell) [inline src/Illuminate/Auth/Passwords/PasswordBroker.php]: We shouldn't inject the entire application here just to get the key. Just inject the key.

Comment (tillkruss) [inline src/Illuminate/Auth/Passwords/PasswordBroker.php]: That makes sense. Changed in https://github.com/laravel/framework/pull/17499/commits/c709b25f6b10e0a247cc46cdbfaf0dc1b7091d52.

**Comment by sisve** [comment]
Would it be possible to keep and deprecate the old DatabaseTokenRepository for a release, instead of completely removing it? 

Relevant: https://github.com/laravel/internals/issues/383 (Deprecating changes for 1 release instead of removing)

**Comment by garygreen** [comment]
@sisve the suggestion in that thread seems to allude towards drepecte things "that can be aliased" - I don't see any point in keeping the old method for one release - having it as upgrade notes should be suitable.

**Comment by tillkruss** [comment]
As an alternative to deprecating it, the `DatabaseTokenRepository` could be move into a 3rd-party package, just like `tinker` and `browser-kit-testing` was in v5.4.

**Comment by browner12** [comment]
can you ELI5 what this means, and what the process kind of looks like?

**Comment by molteber** [comment]
Is this somewhat simular as a JWT? The concept looks familiar. Looks good :+1:

**Comment by tillkruss** [comment]
Yeah somewhat similar, [here is an article on it](https://neosmart.net/blog/2015/using-hmac-signatures-to-avoid-database-writes/).

**Comment by sisve** [comment]
@garygreen Taylor has now [commented](https://github.com/laravel/internals/issues/383#issuecomment-286576616) that "will attempt to deprecate removals for 1 release".

I see no reason that crypto-based password resets cannot live side-by-side. Isn't it, to over-simplify it, just two different kinds of token generators/validators; one that uses a repository, another that uses this new technology?

Or, to turn the question around; why must the old functionality be removed when this is introduced?

**Comment by taylorotwell** [comment]
Honestly I think I may table this for now unless there is a severe security problem with our current approach. Trying to stabilize out these releases a bit more.

**Comment by tillkruss** [comment]
Okay. If anyone wants this a package, LMK.
